(chore): updating and cleaning parametric.md (#3210)
DataDog/system-tests: docs/scenarios/parametric.md

- Docker
- Python 3.12

then, run the following command, which will create a Python virtual environment and install the Python dependencies from the root directory:

```sh

> ./build.sh -i runner

Reviewer: Why removing this?

Author: The run command automatically builds so I thought it was not needed. Let me know if you think I should add it back!

Reviewer: the run command only build the tested container. `./build.sh -i runner` builds the python virtual env used in ./run.sh`, so it's needed. Though, you need it only once, until we don't update system-tests requirements.

Author: I'll add it back! Thank you

Tests can be aborted using CTRL-C but note that containers maybe still be running and will have to be shut down.

### Using Pytest

The tests are executed using pytest. Below are some common command-line options you can use to control and customize your test runs.
- `-k EXPRESSION`: Run tests that match the given expression (substring or pattern). Useful for running specific tests or groups of tests.

```sh
TEST_LIBRARY=dotnet ./run.sh PARAMETRIC -k test_metrics_msgpack_serialization_TS001
```

- `-L`: To specifiy a language using an argument rather than env
```sh
./run.sh PARAMETRIC -L dotnet -k test_metrics_msgpack_serialization_TS001
```

- `-v`: Increase verbosity. Shows each test name and its result (pass/fail) as they are run.

```sh
TEST_LIBRARY=dotnet ./run.sh PARAMETRIC -v
```

- `-vv`: Even more verbose output. Provides detailed information including setup and teardown for each test.

```sh
TEST_LIBRARY=dotnet ./run.sh PARAMETRIC -vv -k test_metrics_
```

- `-s`: Disable output capture. Allows you to see print statements and logs directly in the console.

```sh
TEST_LIBRARY=dotnet ./run.sh PARAMETRIC -s
```

> ### Running the tests for a custom tracer

Reviewer 2: Let's move this to either right below or right above ### Running the tests

> To run tests against custom tracers, refer to the [Binaries Documentation](system-tests/docs/execute/binaries.md)

Reviewer 2: I think this link path may not work? Locally for me it jumps to docs/scenarios/system-tests/docs/execute/binaries.md which isn't a real file. I guess you need to use relative path instead, so `../execute/binaries.md' I think?

> #### After Testing with a Custom Tracer:

Reviewer 2: Maybe link to the binaries section?

Tests can be aborted using CTRL-C but note that containers maybe still be running and will have to be shut down.

### Using Pytest

The tests are executed using pytest. Below are some common command-line options you can use to control and customize your test runs.
- `-k EXPRESSION`: Run tests that match the given expression (substring or pattern). Useful for running specific tests or groups of tests.

```sh

> TEST_LIBRARY=dotnet ./run.sh PARAMETRIC -k test_metrics_msgpack_serialization_TS001

Reviewer: just a comment FYI, you can handle it in this PR, or in a later PR, or never :) 

The usual way to specify the lang is by using the `-L` argument : 

```suggestion
./run.sh PARAMETRIC -L dotnet -k test_metrics_msgpack_serialization_TS001
```

It's faster to type. The env var works fine but is longer to type. It;s commonly used for someone who **always** use only one library., and set this in it's `.env` file.

```

- `-v`: Increase verbosity. Shows each test name and its result (pass/fail) as they are run.

```sh
TEST_LIBRARY=dotnet ./run.sh PARAMETRIC -v
```

- `-vv`: Even more verbose output. Provides detailed information including setup and teardown for each test.

```sh
TEST_LIBRARY=dotnet ./run.sh PARAMETRIC -vv -k test_metrics_
```

- `-s`: Disable output capture. Allows you to see print statements and logs directly in the console.

```sh
TEST_LIBRARY=dotnet ./run.sh PARAMETRIC -s
```

> ### Running the tests for a custom tracer

Reviewer 2: So we have this binaries.md file which talks about how to build custom tracers for each language (learned about this recently): https://github.com/DataDog/system-tests/blob/main/docs/execute/binaries.md?plain=1

Could we modify the necessary pieces of binaries.md for each tracer (basically just adding anything that we have in parametric.md regarding this) and then link to binaries.md sections for each custom tracer part? That way the information is just kept in a single place.

#### Python

To run the Python tests against a custom tracer:
```bash
echo “ddtrace @ git+https://github.com/DataDog/dd-trace-py.git@<name-of-your-branch>” > binaries/python-load-from-pip
```

#### NodeJS

> There is three ways for running the NodeJS tests with a custom tracer:

Reviewer 2: ```suggestion
There are three ways for running the NodeJS tests with a custom tracer:
```

    * `https://github.com/DataDog/dd-trace-cpp@<COMMIT HASH>`
2. Clone the dd-trace-cpp repo inside `binaries`

> #### When you are done testing against a custom tracer:

Reviewer 2: Let's link to this section at the top of the `### Running the tests for a custom tracer` section so that new users (the ones most likely to get caught off guard) have a higher chance of seeing this. Something like "note that any modification to the binaries will change the version of the tracer used across your local computer. After you're done testing with the custom version, be sure to remove it from the binaries by following <link>". 

I think you don't need to add an example per tracer here but one or two would be helpful along with just an explanation that you should remove it from the binaries if that's how you added it.

```bash
rm -rf binaries/python-load-from-pip
```

> ### Understanding the test outcomes

Reviewer: Cold you add only a ling to `docs/execute/test-outcomes.md` here ? I'd prefer not to duplicate information in the doc.

Then you should have updated proto files. This script will generate weird files, you can ignore/delete these.

### Architecture: How System-tests work

> Below is an overview of how the  testing architecture is structured:

Reviewer 2: ```suggestion
Below is an overview of how the testing architecture is structured:
```

The Python implementation of the interface `app/python`, when run, provides a specification of the API when run.
See the steps below in the HTTP section to run the Python server and view the specification.

### Shared Interface

#### HTTP

> We have transitioned to using an HTTP interface, replacing the legacy GRPC interface. To view the HTTP interface, follow these steps:

Reviewer 2: ```suggestion
We have transitioned to using an HTTP interface, replacing the legacy GRPC interface. To view the available HTTP endpoints , follow these steps:
```

Then you should have updated proto files. This script will generate weird files, you can ignore/delete these.

> ### Architecture

Reviewer 2: ```suggestion
### Architecture/How System-tests work
```
Maybe it's good to add this or replace architecture with it? I just want it to be obvious to new contributors that they should probably look at this to understand what's going on.

Reviewer 2: If you think that new people would be as likely to read an "Architecture" section as my suggested change then feel free to ignore.

Below is an overview of how the  testing architecture is structured:

> - Shared Tests in Python: We write shared test cases using Python's pytest framework. These tests are designed to be generic and interact with clients through the HTTP interface.

Reviewer 2: ```suggestion
- Shared Tests in Python: We write shared test cases using Python's pytest framework. These tests are designed to be generic and interact with the tracers through an HTTP interface.
```

Reviewer 2: Can we maybe use the word "tracer" instead of "client" here and below? I feel like client is so obscure.

Reviewer 2: In general, I just want users to understand that the methods written in test code to generate trace data make an outgoing http request through an Http client to an http server. On the http server the tracer is running and we carry out various operations with the tracer (e.g. generating spans, or adding tags) depending on which endpoints you hit and what values are passed in to the outgoing request.

From there we have two ways to check that the logic was carried out as expected, getting spans from the testagent (should provide example probably?) or looking at the value returned by the http server (example would also probably be helpful.)
